File: Demos/Todos/ObjectiveC/Todos/TableViewDataSource.m
//
//  TableViewDataSource.m
//  Toodoos
//
//  Created by Sasa Prodribaba on 13/12/2018.
//  Copyright © 2018 Sasa Prodribaba. All rights reserved.
//

#import "TableViewDataSource.h"
#import "Fetcher.h"
#import "Todo.h"

@implementation TableViewDataSource

#define kTodoCellIdentifier @"todoCell"

NSArray *todos;

- (instancetype)init {
    self = [super init];
    if (self) {
        todos = [NSArray new];
        Fetcher *fetcher = [Fetcher new];
        [fetcher fetchWithResult:^(NSArray *result) {
            todos = result;
            [self.dataSourceDelegate didReceiveData];
        }];
    }
    return self;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kTodoCellIdentifier forIndexPath:indexPath];
    cell.textLabel.text = ((Todo *)todos[indexPath.row]).title;
    cell.detailTextLabel.text = [((Todo *)todos[indexPath.row]).completed boolValue]?@"Completed":@"Due";
    
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [todos count];
}

@end

File: Demos/Todos/Swift/Todos/TableViewDataSource.swift
//
//  TableViewDataSource.swift
//  Todos
//
//  Created by Sasa Prodribaba on 25/12/2018.
//  Copyright © 2018 Sasa Prodribaba. All rights reserved.
//

import UIKit

protocol DataSourceDelegate {
    func didReceiveData()
}

class TableViewDataSource: NSObject, UITableViewDataSource {
    
    var dataSourceDelegate: DataSourceDelegate?
    
    var todos = [Todo]()
    
    override init() {
        super.init()
        
        let fetcher = Fetcher()
        fetcher.fetchWithResult {
            result in
            self.todos = result
            self.dataSourceDelegate?.didReceiveData()
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: Constants.todoCellIdentifier, for: indexPath)
        cell.textLabel?.text = todos[indexPath.row].title
        cell.detailTextLabel?.text = "\(todos[indexPath.row].completed ? "Completed" : "Due")"
        return cell
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return todos.count
    }
    
}
